File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/c3bcbeddc568924dca3333ae8d7a77d9.py
import sys

def dijkstra(graph, start):
    n = len(graph)
    dist = [float('inf')] * n
    dist[start] = 0
    visited = [False] * n

    for _ in range(n):
        min_dist = float('inf')
        min_node = -1

        for i in range(n):
            if not visited[i] and dist[i] < min_dist:
                min_dist = dist[i]
                min_node = i

        if min_node == -1:
            break

        visited[min_node] = True

        for neighbor, weight in graph[min_node]:
            if dist[min_node] + weight < dist[neighbor]:
                dist[neighbor] = dist[min_node] + weight

    return dist

def solve(n, k, arrs):
    sums = [sum(arr) for arr in arrs]
    dp = [[0] * (k + 1) for _ in range(13)]

    def copy_dp(dpi):
        for i in range(k + 1):
            dp[dpi + 1][i] = dp[dpi][i]

    def add_to_knap(i, dpi):
        v = sums[i]
        k1 = len(arrs[i])
        for j in range(k, k1 - 1, -1):
            dp[dpi][j] = max(dp[dpi][j], dp[dpi][j - k1] + v)

    def rec(l, r, dpi):
        ans = 0
        carr = arrs[l]
        if l + 1 == r:
            ans = max(ans, dp[dpi][k])
            s = 0
            for i in range(min(k, len(carr))):
                s += carr[i]
                ans = max(ans, dp[dpi][k - i - 1] + s)
        else:
            m = (l + r) // 2
            copy_dp(dpi)
            for i in range(l, m):
                add_to_knap(i, dpi + 1)
            ans = max(ans, rec(m, r, dpi + 1))
            copy_dp(dpi)
            for i in range(m, r):
                add_to_knap(i, dpi + 1)
            ans = max(ans, rec(l, m, dpi + 1))
        return ans

    return rec(0, n, 0)

def main():
    n, k = map(int, input().split())
    arrs = []
    for _ in range(n):
        t = int(input())
        arr = list(map(int, input().split()))
        if t > k:
            arr = arr[:k]
        arrs.append(arr)
    print(solve(n, k, arrs))

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/c3bcbeddc568924dca3333ae8d7a77d9.java
import java.io.*;
import java.util.*;
import java.util.function.Function;

public class c3bcbeddc568924dca3333ae8d7a77d9_nc {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;
        private int k;
        private long[][] dps;
        private long[] sums;
        private List<List<Long>> arrs;

        Solver(InputStream input, OutputStream output) {
            in = new BufferedReader(new InputStreamReader(input));
            out = new PrintWriter(output);
        }

        private void readInt(int[] args) throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            for (int i = 0; i < args.length; i++) {
                args[i] = Integer.parseInt(st.nextToken());
            }
        }

        private void readLong(long[] args) throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            for (int i = 0; i < args.length; i++) {
                args[i] = Long.parseLong(st.nextToken());
            }
        }

        private void doCopy(int dpi) {
            System.arraycopy(dps[dpi], 0, dps[dpi + 1], 0, dps[0].length);
        }

        private void addToKnap(int i, int dpi) {
            long v = sums[i];
            int k1 = arrs.get(i).size();
            for (int j = k; j >= k1; j--) {
                dps[dpi][j] = Math.max(dps[dpi][j], dps[dpi][j - k1] + v);
            }
        }

        private long rec(int l, int r, int dpi) {
            long ans = 0;
            List<Long> carr = arrs.get(l);
            long[] dp = dps[dpi];

            if (l + 1 == r) {
                ans = Math.max(ans, dp[k]);
                long sum = 0;
                for (int i = 0; i < Math.min(k, carr.size()); i++) {
                    sum += carr.get(i);
                    ans = Math.max(ans, dp[k - i - 1] + sum);
                }
            } else {
                int m = (l + r) / 2;
                doCopy(dpi);
                for (int i = l; i < m; i++) {
                    addToKnap(i, dpi + 1);
                }
                ans = Math.max(ans, rec(m, r, dpi + 1));
                doCopy(dpi);
                for (int i = m; i < r; i++) {
                    addToKnap(i, dpi + 1);
                }
                ans = Math.max(ans, rec(l, m, dpi + 1));
            }
            return ans;
        }

        void solve() throws IOException {
            int n = Integer.parseInt(in.readLine());
            k = Integer.parseInt(in.readLine());
            sums = new long[n];
            arrs = new ArrayList<>(n);
            for (int i = 0; i < n; i++) {
                String[] line = in.readLine().split(" ");
                int t = Integer.parseInt(line[0]);
                List<Long> arr = new ArrayList<>();
                for (int j = 1; j <= t; j++) {
                    arr.add(Long.parseLong(line[j]));
                }
                if (t > k) {
                    arr = arr.subList(0, k);
                }
                arrs.add(arr);
                sums[i] = arr.stream().mapToLong(Long::longValue).sum();
            }
            dps = new long[13][k + 1];
            out.println(rec(0, n, 0));
        }
    }

    public static void main(String[] args) throws IOException {
        DSum solver = new DSum();
        solver.solve(System.in, System.out);
    }
}